buildSrc/subprojects/parser/src/main/kotlin/org/apache/calcite/buildtools/parser/DialectGenerateTask.kt

```diff
@@ -127,25 +146,74 @@ open class DialectGenerateTask @Inject constructor(
      */
     private fun processFile(file: File, functionMap: MutableMap<String, String>) {
         var fileText = file.readText(Charsets.UTF_8)
-        val typeAndName = "\\w+\\s+\\w+"
-        val declarationPattern =
-            Regex("(%s\\s*\\(\\s*(%s\\s*(\\,\\s*%s\\s*)*)?\\)\\s*\\:\n)"
-                    .format(typeAndName, typeAndName, typeAndName))
-        val delims = "(\\s|\n|\"|(//)|(/\\*)|(\\*/))"
-        // Uses Lookahead and Lookbehind to ensure that the delims are added as tokens.
-        val splitRegex = Regex("((?<=%s)|(?=%s))".format(delims, delims))
-        val declarationMatches = declarationPattern.findAll(fileText)
-        for (m in declarationMatches) {
-            val functionDeclaration = m.value
-            val functionName = getFunctionName(functionDeclaration)
-            val functionBuilder = StringBuilder(functionDeclaration)
-            val declarationIndex = fileText.indexOf(functionDeclaration)
-            fileText = fileText.substring(declarationIndex + functionDeclaration.length)
-            val tokens: Queue<String> = LinkedList(splitRegex.split(fileText))
-            processCurlyBlock(functionBuilder, tokens)
-            processCurlyBlock(functionBuilder, tokens)
-            functionMap.put(functionName, functionBuilder.toString())
+        val declarations: Queue<MatchResult> = LinkedList(declarationPattern.findAll(fileText).toList())
```

Comment:
I think we may be able to simplify the logic in this function a bit. How about this alternative representation? (Disclaimer: there may be bugs :)

val declarations: Queue<MatchResult> = LinkedList(declarationPattern.findAll(fileText).toList())
val tokens: Queue<String> = LinkedList(tokenizer.split(fileText))
var charIndex = 0
while (declarations.isNotEmpty()) {
  declaration = declarations.poll()
  while (charIndex < declaration.range.start) {
    charIndex += tokens.poll().length
  }
  charIndex = processFunction(tokens, functionMap, charIndex,
      declaration.range.endInclusive, getFunctionName(declaration.value))
}

Reply:
Very nice! With a few syntax fixes it work :)